CMakeLists.txt

```diff
@@ -0,0 +1,279 @@
+        " ZYNQ                   Target: Xilinx Zynq")
+elseif((FREERTOS_PORT STREQUAL "A_CUSTOM_PORT") AND (NOT TARGET freertos_kernel_port) )
+    message(FATAL_ERROR " FREERTOS_PLUS_TCP_NETWORK_IF is set to A_CUSTOM_NETWORK_IF.\n"
+    " Please specify the custom network interface target with all necessary files.\n"
+    " For example, assuming a directory of:\n"
+    "  FreeRTOSCustomNetworkInterface/\n"
+    "    CMakeLists.txt\n"
+    "    NetworkInterface.c\n\n"
+    " Where FreeRTOSCustomNetworkInterface/CMakeLists.txt is a modified version of:\n"
+    "   add_library(freertos_plus_tcp_network_if STATIC)\n\n"
+    "   target_sources(freertos_plus_tcp_network_if\n"
+    "     PRIVATE\n"
+    "       NetworkInterface.c)\n\n"
+    "   target_include_directories(freertos_plus_tcp_network_if\n"
+    "     PUBLIC\n"
+    "      .)\n\n"
+    "   taget_link_libraries(freertos_plus_tcp_network_if\n"
+    "     PRIVATE\n"
+    "       freertos_kernel)")
+endif()
+
+# There is also the need to add a target - typically an interface library that describes the
+# Configuration for FreeRTOS-Kernel and FreeRTOS-Plus-TCP.
+# This is called FreeRTOS::Config
+# If not defined will be default compile with one of the test build combinations AllEnable.
+
+# Select the appropriate Build Test configuration
+# This is only used when FreeRTOS::Config is not defined, otherwise the build test will be performed
+# on the config defined in the FreeRTOS::Config
+set(FREERTOS_PLUS_TCP_TEST_CONFIGURATION "DEFAULT_CONF" CACHE STRING "FreeRTOS Plus TCP Build Test configuration")
+set(FREERTOS_PLUS_TCP_TEST_CONFIGURATION_LIST
+    CUSTOM       # Custom (external) configuration -eg from a top-level project
+    ENABLE_ALL   # Enable all configuration settings
+    DISABLE_ALL  # Disable all configuration settings
+    DEFAULT_CONF # Default (typical) configuration)
+)
+if(NOT FREERTOS_PLUS_TCP_TEST_CONFIGURATION IN_LIST FREERTOS_PLUS_TCP_TEST_CONFIGURATION_LIST)
+    message(FATAL_ERROR "Invalid FREERTOS_PLUS_TCP_TEST_CONFIGURATION value '${FREERTOS_PLUS_TCP_TEST_CONFIGURATION}' should be one of: ${FREERTOS_PLUS_TCP_TEST_CONFIGURATION_LIST}")
+else()
+    message(WARNING "Using FreeRTOS-Plus-TCP Test Configuration : ${FREERTOS_PLUS_TCP_TEST_CONFIGURATION}"
+    "  FreeRTOS-Kernel configuration settings are configured by FreeRTOS-Plus-TCP")
+    if (NOT FREERTOS_PLUS_TCP_TEST_CONFIGURATION STREQUAL "CUSTOM")
+        # Configuration for FreeRTOS-Kernel
+        set(FREERTOS_CONFIG_FILE_DIRECTORY "${CMAKE_CURRENT_LIST_DIR}/test/build-combination/Common" CACHE STRING "" FORCE)
+        if(UNIX)
+            set(FREERTOS_PORT GCC_POSIX CACHE STRING "" FORCE)
+        elseif(MINGW)
+            set(FREERTOS_PORT MSVC_MINGW CACHE STRING "" FORCE)
+        endif()
+    endif()
+endif()
+
+########################################################################
+# Requirements
+set(CMAKE_C_STANDARD 99) # Note FreeRTOS-Kernel uses C99 constructs.
```

Comment:
The kernel should compile with C89

Reply:
@archigup - IMO, that ship has sailed.  Unless there is a significant amount of work to remove all of the code that is using C99 constructs and operation is done in both the FreeRTOS-Kernel and FreeRTOS-Plus-TCP this is not viable at the moment.

One example is `declaration after execution` - i.e. declaring variables midway in a procedure.  There are many cases.

Is there a particular compiler that doesn't support C99 at this point in time? - For those users who wish to continue compiling with a compiler with no C99 support, I believe if they are not likely to update their compiler, what are the chances they will consume a new version of FreeRTOS-Plus-TCP as well?  We are talking about 20+ yr old compilers here.

Reply:
@phelter : We have had a number of internal discussions on this topic and our current policy is ANSI C89 / ISO C90 compliance with a limited number of exceptions (stdbool and stdint mainly). I don't expect this policy to change in the near future until we can be 100% certain there are no compatibility breaks.

If there are specific C89/C90 compliance problems, they should be either:
1) Considered a bug <OR>
2) Added to the list of exceptions for that particular repository in the repository's README.md and masked for common compiler in the cmake or source files.

Note, this mainly applies to library code. I'ts a bit more flexible with demos or other examples that target a particular platform. In the case of FreeRTOS+TCP, the portable directory is excluded from this policy.

Reply:
Ok - thanks - will assume all use of C99 as a bug then and identify them in the warnings for now - and fix them in a follow-on commit.

Reply:
FYI:

I've fixed all the C90 - ISO violations in the code and have separated out the cmakes so they are per Network IF port now instead of all one big lump with multiple `case` statements in the one.

I do not intend to test all of the other ports but have tested the linux and Zynq ports and have confirmed they work.

Reply:
@paulbartell - One other thing to note.  If you had example builds of each and every compiler that FreeRTOS-Plus-TCP (and FreeRTOS-Kernel) intends to support, then you could use those as a deciding factor as to whether you could increase to C99 or higher (eg C11).  The example builds could be easily provided in a CI/CD build by:
* Creating a docker image (eg. Linux Ubuntu) per compiler supported - along with installed BSP of each of the boards that is identified as supported.
* Provide a toolchain file per compiler
* Provide a toolchain file per board (optional) that includes the specific compile options for that board and includes the compiler toolchain (Dont-Repeat-Yourself DRY).  This is optional because if you only want to confirm the compiler works then you really only need one example and all others should be the same with the exception of the portable code that is compiled.  To reduce the amount of time checking all the different boards you could compile everything with one board BSP and then just compile the `freertos_plus_tcp_network_if` target for each of the other boards supported with that same compiler.
* Add the necessary build as a cross-compilation per compiler and board in the CI/CD flow - probably only necessary for release builds.

Then there is a guarantee that if you change something like the Required Compiler ISO - you know it won't break someone's build.